File: audio-device/swift/Sources/AudioChatDeviceKit/AudioChatDevice.swift
import Foundation

public struct AudioChatDevice {
    public var deviceID: String
    public var userID: String
    public var name: String
    public var role: String?
    public var runtime: [String: Any]
    public var properties: [String: Any]
    public var sensors: [[String: Any]]
    public var actuators: [[String: Any]]

    public init(deviceID: String) {
        self.deviceID = deviceID
        self.userID = ""
        self.name = deviceID
        self.runtime = ["platform": "ios", "language": "swift"]
        self.properties = [:]
        self.sensors = []
        self.actuators = []
    }

    public func user(_ userID: String) -> AudioChatDevice {
        var copy = self
        copy.userID = userID
        return copy
    }

    public func named(_ name: String) -> AudioChatDevice {
        var copy = self
        copy.name = name
        return copy
    }

    public func role(_ role: String) -> AudioChatDevice {
        var copy = self
        copy.role = role
        return copy
    }

    public func sensorRgb(modes: [String] = ["single"], format: String = "jpeg", frequencyHz: Double? = nil) -> AudioChatDevice {
        var copy = self
        var defaults: [String: Any] = ["format": format]
        if let frequencyHz { defaults["frequency_hz"] = frequencyHz }
        copy.sensors.append(["type": "rgb", "modes": modes, "default": defaults])
        return copy
    }

    public func actuatorVibrator(commands: [String] = ["vibrate"]) -> AudioChatDevice {
        var copy = self
        copy.actuators.append(["type": "vibrator", "commands": commands])
        return copy
    }

    public var registrationPayload: [String: Any] {
        var props = properties
        if let role { props["device_role"] = role }
        var supports: [String: Any] = [:]
        if !sensors.isEmpty { supports["sensors"] = sensors }
        if !actuators.isEmpty { supports["actuators"] = actuators }
        return [
            "device_id": deviceID,
            "name": name,
            "device_name": name,
            "client_type": runtime["platform"] as? String ?? "ios",
            "sdk_version": "0.1.0",
            "runtime": runtime,
            "properties": props,
            "supports": supports,
        ]
    }
}

File: audio-device/kotlin/device/src/main/kotlin/io/audiochat/device/AudioChatDevice.kt
package io.audiochat.device

/**
 * 设备声明构建器
 * 用于生成设备注册 payload
 */
class AudioChatDevice private constructor(val device_id: String) {
    private var user_id: String = ""
    var name: String = device_id
        private set
    private var role: String? = null
    private var runtime: Map<String, Any?> = mapOf("platform" to "unknown", "language" to "kotlin")
    private val properties = linkedMapOf<String, Any?>()
    private val sensors = mutableListOf<Map<String, Any?>>()
    private val actuators = mutableListOf<Map<String, Any?>>()

    companion object {
        @JvmStatic
        fun define(deviceId: String): AudioChatDevice = AudioChatDevice(deviceId)
    }

    fun user(userId: String): AudioChatDevice {
        this.user_id = userId
        return this
    }

    fun name(name: String): AudioChatDevice {
        this.name = name
        return this
    }

    fun role(role: String): AudioChatDevice {
        this.role = role
        return this
    }

    fun platform(platform: String): AudioChatDevice {
        this.runtime = this.runtime.toMutableMap().apply { put("platform", platform) }
        return this
    }

    fun property(key: String, value: Any?): AudioChatDevice {
        properties[key] = value
        return this
    }

    fun sensorRgb(
        modes: List<String> = listOf("single"),
        format: String = "jpeg",
        frequencyHz: Int? = null
    ): AudioChatDevice {
        val defaults = linkedMapOf<String, Any?>("format" to format)
        if (frequencyHz != null) defaults["frequency_hz"] = frequencyHz
        sensors.add(mapOf("type" to "rgb", "modes" to modes, "default" to defaults))
        return this
    }

    fun sensorMic(
        modes: List<String> = listOf("continuous"),
        sampleRate: Int = 16000,
        channels: Int = 1
    ): AudioChatDevice {
        sensors.add(mapOf(
            "type" to "mic",
            "modes" to modes,
            "default" to mapOf(
                "sample_rate" to sampleRate,
                "channels" to channels,
                "codec" to "pcm16le"
            )
        ))
        return this
    }

    fun actuatorSpeaker(
        modes: List<String> = listOf("continuous"),
        sampleRate: Int = 16000,
        channels: Int = 1
    ): AudioChatDevice {
        actuators.add(mapOf(
            "type" to "speaker",
            "modes" to modes,
            "default" to mapOf(
                "sample_rate" to sampleRate,
                "channels" to channels,
                "codec" to "pcm16le"
            )
        ))
        return this
    }

    fun actuatorVibrator(commands: List<String> = listOf("vibrate")): AudioChatDevice {
        actuators.add(mapOf("type" to "vibrator", "commands" to commands))
        return this
    }

    fun supports(supports: DeviceSupports): AudioChatDevice {
        supports.sensors.forEach { sensors.add(it.toMap()) }
        supports.actuators.forEach { actuators.add(it.toMap()) }
        return this
    }

    fun registrationPayload(): Map<String, Any?> {
        require(user_id.isNotBlank()) { "user_id is required" }
        val props = linkedMapOf<String, Any?>()
        props.putAll(properties)
        if (role != null) props["device_role"] = role

        val supportsMap = linkedMapOf<String, Any?>()
        if (sensors.isNotEmpty()) supportsMap["sensors"] = sensors
        if (actuators.isNotEmpty()) supportsMap["actuators"] = actuators

        return linkedMapOf(
            "device_id" to device_id,
            "name" to name,
            "device_name" to name,
            "client_type" to (runtime["platform"] ?: "unknown"),
            "sdk_version" to "1.0.0",
            "runtime" to runtime,
            "properties" to props,
            "supports" to supportsMap
        )
    }

    /**
     * 创建手机设备
     */
    fun asPhone(): AudioChatDevice {
        return this
            .role(DeviceConstants.DEVICE_TYPE_PHONE)
            .sensorMic()
            .actuatorSpeaker()
    }

    /**
     * 创建眼镜设备
     */
    fun asGlass(): AudioChatDevice {
        return this
            .role(DeviceConstants.DEVICE_TYPE_GLASS)
            .sensorRgb()
    }
}